{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "df26c46a",
   "metadata": {},
   "source": [
    "Purpose: See if any of my core genes overlap with GWAS hits (from paper: https://academic-oup-com.proxy1.cl.msu.edu/gigascience/article/doi/10.1093/gigascience/giac080/6673780#369569254).<br>\n",
    "Author: Anna Pardo<br>\n",
    "Date initiated: Sept. 5, 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1a27ed0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import modules\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import fdrcorrection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08c786fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peakName</th>\n",
       "      <th>PeakChromosome</th>\n",
       "      <th>PeakSNPPosition</th>\n",
       "      <th>MaximumRMIPScore</th>\n",
       "      <th>SNPsInPeak</th>\n",
       "      <th>PeakStart</th>\n",
       "      <th>PeakEnd</th>\n",
       "      <th>PeakLength</th>\n",
       "      <th>ClosestGeneToPeak</th>\n",
       "      <th>SecondClosestGeneToPeak</th>\n",
       "      <th>ThirdClosestGeneToPeak</th>\n",
       "      <th>NumberOfSignificantTraits</th>\n",
       "      <th>AllSignificantTraits</th>\n",
       "      <th>NumberOfSignificantPhenotypeGroups</th>\n",
       "      <th>PhenotypeGroups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>160,809,058</td>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>160,559,294</td>\n",
       "      <td>160,989,691</td>\n",
       "      <td>430397</td>\n",
       "      <td>Zm00001d042317</td>\n",
       "      <td>Zm00001d042319</td>\n",
       "      <td>Zm00001d042315</td>\n",
       "      <td>16</td>\n",
       "      <td>Anthesis1_L;Anthesis4_H;Anthesis6_H;Anthesis7_...</td>\n",
       "      <td>2</td>\n",
       "      <td>FloweringTime;Vegetative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>136,012,710</td>\n",
       "      <td>57</td>\n",
       "      <td>24</td>\n",
       "      <td>135,928,821</td>\n",
       "      <td>136,325,345</td>\n",
       "      <td>396524</td>\n",
       "      <td>Zm00001d010988</td>\n",
       "      <td>Zm00001d010987</td>\n",
       "      <td>Zm00001d010989</td>\n",
       "      <td>12</td>\n",
       "      <td>Anthesis1_L;Anthesis5_H;Anthesis6_H;Anthesis7_...</td>\n",
       "      <td>2</td>\n",
       "      <td>FloweringTime;Vegetative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>113,976,392</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>112,979,061</td>\n",
       "      <td>113,976,392</td>\n",
       "      <td>997331</td>\n",
       "      <td>Zm00001d020433</td>\n",
       "      <td>Zm00001d020432</td>\n",
       "      <td>Zm00001d020434</td>\n",
       "      <td>5</td>\n",
       "      <td>BranchDensity_C;BranchNumber1_C;BranchZoneLeng...</td>\n",
       "      <td>1</td>\n",
       "      <td>Inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>155,019,919</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>155,019,919</td>\n",
       "      <td>155,021,976</td>\n",
       "      <td>2057</td>\n",
       "      <td>Zm00001d011585</td>\n",
       "      <td>Zm00001d011584</td>\n",
       "      <td>Zm00001d011583</td>\n",
       "      <td>3</td>\n",
       "      <td>Anthesis4_H;Anthesis7_H;Anthesis_J</td>\n",
       "      <td>1</td>\n",
       "      <td>FloweringTime</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>15,523,359</td>\n",
       "      <td>57</td>\n",
       "      <td>13</td>\n",
       "      <td>14,881,429</td>\n",
       "      <td>16,102,546</td>\n",
       "      <td>1221117</td>\n",
       "      <td>Zm00001d035227</td>\n",
       "      <td>Zm00001d035228</td>\n",
       "      <td>Zm00001d035229</td>\n",
       "      <td>5</td>\n",
       "      <td>SMV10DAI_B;SMV14DAI_B;SMV21DAI_B;SMV7DAI_B;SMV...</td>\n",
       "      <td>1</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   peakName  PeakChromosome PeakSNPPosition  MaximumRMIPScore  SNPsInPeak  \\\n",
       "0         1               3     160,809,058                56          57   \n",
       "1         2               8     136,012,710                57          24   \n",
       "2         3               7     113,976,392                19          12   \n",
       "3         4               8     155,019,919                17          10   \n",
       "4         5               6      15,523,359                57          13   \n",
       "\n",
       "     PeakStart      PeakEnd  PeakLength ClosestGeneToPeak  \\\n",
       "0  160,559,294  160,989,691      430397    Zm00001d042317   \n",
       "1  135,928,821  136,325,345      396524    Zm00001d010988   \n",
       "2  112,979,061  113,976,392      997331    Zm00001d020433   \n",
       "3  155,019,919  155,021,976        2057    Zm00001d011585   \n",
       "4   14,881,429   16,102,546     1221117    Zm00001d035227   \n",
       "\n",
       "  SecondClosestGeneToPeak ThirdClosestGeneToPeak  NumberOfSignificantTraits  \\\n",
       "0          Zm00001d042319         Zm00001d042315                         16   \n",
       "1          Zm00001d010987         Zm00001d010989                         12   \n",
       "2          Zm00001d020432         Zm00001d020434                          5   \n",
       "3          Zm00001d011584         Zm00001d011583                          3   \n",
       "4          Zm00001d035228         Zm00001d035229                          5   \n",
       "\n",
       "                                AllSignificantTraits  \\\n",
       "0  Anthesis1_L;Anthesis4_H;Anthesis6_H;Anthesis7_...   \n",
       "1  Anthesis1_L;Anthesis5_H;Anthesis6_H;Anthesis7_...   \n",
       "2  BranchDensity_C;BranchNumber1_C;BranchZoneLeng...   \n",
       "3                 Anthesis4_H;Anthesis7_H;Anthesis_J   \n",
       "4  SMV10DAI_B;SMV14DAI_B;SMV21DAI_B;SMV7DAI_B;SMV...   \n",
       "\n",
       "   NumberOfSignificantPhenotypeGroups           PhenotypeGroups  \n",
       "0                                   2  FloweringTime;Vegetative  \n",
       "1                                   2  FloweringTime;Vegetative  \n",
       "2                                   1             Inflorescence  \n",
       "3                                   1             FloweringTime  \n",
       "4                                   1                   Disease  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load GWAS table (supplementary table S4 from paper linked above)\n",
    "gwas = pd.read_csv(\"../../data/SupplementalDataFileS4_AllUniqueGWASPeaks.csv\",sep=\",\",header=\"infer\")\n",
    "gwas.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8c106a60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# these gene IDs are from B73 V4\n",
    "## make a list of genes to paste into Ensembl's ID History Converter\n",
    "## or just write them to a file...but first make a non-overlapping set (set union of all these genes)\n",
    "c = set(list(gwas[\"ClosestGeneToPeak\"]))\n",
    "sc = set(list(gwas[\"SecondClosestGeneToPeak\"]))\n",
    "tc = set(list(gwas[\"ThirdClosestGeneToPeak\"]))\n",
    "allc = set.union(*[c,sc,tc])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d4af034a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"../../data/v4_genes_from_Addie_GWAS_GigaScience.txt\",\"w+\") as outfile:\n",
    "    for g in allc:\n",
    "        outfile.write(g+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6242c482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID_V5</th>\n",
       "      <th>GeneID_V4</th>\n",
       "      <th>Second_GeneID_V4</th>\n",
       "      <th>Unnamed: 3</th>\n",
       "      <th>Unnamed: 4</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001eb000750</td>\n",
       "      <td>Zm00001d027320</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001eb001060</td>\n",
       "      <td>Zm00001d027353</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001eb001240</td>\n",
       "      <td>Zm00001d027371</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001eb002180</td>\n",
       "      <td>Zm00001d027480</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001eb002270</td>\n",
       "      <td>Zm00001d027488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         GeneID_V5       GeneID_V4 Second_GeneID_V4 Unnamed: 3 Unnamed: 4  \\\n",
       "0  Zm00001eb000750  Zm00001d027320              NaN        NaN        NaN   \n",
       "1  Zm00001eb001060  Zm00001d027353              NaN        NaN        NaN   \n",
       "2  Zm00001eb001240  Zm00001d027371              NaN        NaN        NaN   \n",
       "3  Zm00001eb002180  Zm00001d027480              NaN        NaN        NaN   \n",
       "4  Zm00001eb002270  Zm00001d027488              NaN        NaN        NaN   \n",
       "\n",
       "  Unnamed: 5  \n",
       "0        NaN  \n",
       "1        NaN  \n",
       "2        NaN  \n",
       "3        NaN  \n",
       "4        NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load list of core gene IDs as both V5 and V4 genes\n",
    "cg = pd.read_csv(\"../../data/All_coregenes_V5-to-V4.txt\",sep=\"\\t\",header=\"infer\")\n",
    "cg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "454d2070",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "yes\n"
     ]
    }
   ],
   "source": [
    "if type(cg.iloc[0,2])==float:\n",
    "    print(\"yes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b3f5acb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert this into a dictionary format\n",
    "cgdict = {}\n",
    "for g in range(len(list(cg[\"GeneID_V5\"]))):\n",
    "    k = cg.iloc[g,0]\n",
    "    v4list = []\n",
    "    for i in range(1,6):\n",
    "        if type(cg.iloc[g,i])==str:\n",
    "            v4list.append(cg.iloc[g,i])\n",
    "    cgdict[k]=v4list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6a46e9dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the dataframe of peaks where the closest to peak is a core gene\n",
    "c_v4 = []\n",
    "sc_v4 = []\n",
    "tc_v4 = []\n",
    "cv5 = []\n",
    "scv5 = []\n",
    "tcv5 = []\n",
    "for k in cgdict.keys():\n",
    "    l = cgdict[k]\n",
    "    for g in l:\n",
    "        if g in list(gwas[\"ClosestGeneToPeak\"].unique()):\n",
    "            c_v4.append(g)\n",
    "            cv5.append(k)\n",
    "        if g in list(gwas[\"SecondClosestGeneToPeak\"].unique()):\n",
    "            sc_v4.append(g)\n",
    "            scv5.append(k)\n",
    "        if g in list(gwas[\"ThirdClosestGeneToPeak\"].unique()):\n",
    "            tc_v4.append(g)\n",
    "            tcv5.append(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f9a3dece",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make dataframes\n",
    "cdf = pd.DataFrame(list(zip(cv5,c_v4)),columns=[\"Closest_V5\",\"ClosestGeneToPeak\"])\n",
    "scdf = pd.DataFrame(list(zip(scv5,sc_v4)),columns=[\"SecondClosest_V5\",\"SecondClosestGeneToPeak\"])\n",
    "tcdf = pd.DataFrame(list(zip(tcv5,tc_v4)),columns=[\"ThirdClosest_V5\",\"ThirdClosestGeneToPeak\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cb0a8d02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cdf.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3d5c4016",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(scdf.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2d02759f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tcdf.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9d7aaa4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge with GWAS results to get the agronomic traits linked with the closest genes\n",
    "gwas_cdf = cdf.merge(gwas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2ee12059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Closest_V5</th>\n",
       "      <th>ClosestGeneToPeak</th>\n",
       "      <th>peakName</th>\n",
       "      <th>PeakChromosome</th>\n",
       "      <th>PeakSNPPosition</th>\n",
       "      <th>MaximumRMIPScore</th>\n",
       "      <th>SNPsInPeak</th>\n",
       "      <th>PeakStart</th>\n",
       "      <th>PeakEnd</th>\n",
       "      <th>PeakLength</th>\n",
       "      <th>SecondClosestGeneToPeak</th>\n",
       "      <th>ThirdClosestGeneToPeak</th>\n",
       "      <th>NumberOfSignificantTraits</th>\n",
       "      <th>AllSignificantTraits</th>\n",
       "      <th>NumberOfSignificantPhenotypeGroups</th>\n",
       "      <th>PhenotypeGroups</th>\n",
       "      <th>ClosestCore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001eb000750</td>\n",
       "      <td>Zm00001d027320</td>\n",
       "      <td>290</td>\n",
       "      <td>1</td>\n",
       "      <td>2,882,752</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2,882,752</td>\n",
       "      <td>2,882,752</td>\n",
       "      <td>0</td>\n",
       "      <td>Zm00001d027321</td>\n",
       "      <td>Zm00001d027319</td>\n",
       "      <td>1</td>\n",
       "      <td>BushelAcreEquivalent_J</td>\n",
       "      <td>1</td>\n",
       "      <td>Agronomic</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001eb006310</td>\n",
       "      <td>Zm00001d027936</td>\n",
       "      <td>181</td>\n",
       "      <td>1</td>\n",
       "      <td>17,966,847</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>17,966,847</td>\n",
       "      <td>17,966,847</td>\n",
       "      <td>0</td>\n",
       "      <td>Zm00001d027934</td>\n",
       "      <td>Zm00001d027935</td>\n",
       "      <td>1</td>\n",
       "      <td>NodesWithBraceRoots_J</td>\n",
       "      <td>1</td>\n",
       "      <td>Root</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001eb021040</td>\n",
       "      <td>Zm00001d029628</td>\n",
       "      <td>357</td>\n",
       "      <td>1</td>\n",
       "      <td>80,602,342</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>80,602,342</td>\n",
       "      <td>80,602,342</td>\n",
       "      <td>0</td>\n",
       "      <td>Zm00001d029629</td>\n",
       "      <td>Zm00001d029627</td>\n",
       "      <td>1</td>\n",
       "      <td>EarWidth_J</td>\n",
       "      <td>1</td>\n",
       "      <td>Inflorescence</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001eb024820</td>\n",
       "      <td>Zm00001d030086</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>104,530,524</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>104,530,524</td>\n",
       "      <td>104,530,524</td>\n",
       "      <td>0</td>\n",
       "      <td>Zm00001d030087</td>\n",
       "      <td>Zm00001d030088</td>\n",
       "      <td>1</td>\n",
       "      <td>EarsPerPlant_J</td>\n",
       "      <td>1</td>\n",
       "      <td>Agronomic</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001eb047280</td>\n",
       "      <td>Zm00001d032873</td>\n",
       "      <td>225</td>\n",
       "      <td>1</td>\n",
       "      <td>241,078,399</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>241,078,399</td>\n",
       "      <td>241,078,399</td>\n",
       "      <td>0</td>\n",
       "      <td>Zm00001d032874</td>\n",
       "      <td>Zm00001d032870</td>\n",
       "      <td>1</td>\n",
       "      <td>RootAngle_I</td>\n",
       "      <td>1</td>\n",
       "      <td>Root</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Closest_V5 ClosestGeneToPeak  peakName  PeakChromosome  \\\n",
       "0  Zm00001eb000750    Zm00001d027320       290               1   \n",
       "1  Zm00001eb006310    Zm00001d027936       181               1   \n",
       "2  Zm00001eb021040    Zm00001d029628       357               1   \n",
       "3  Zm00001eb024820    Zm00001d030086       148               1   \n",
       "4  Zm00001eb047280    Zm00001d032873       225               1   \n",
       "\n",
       "  PeakSNPPosition  MaximumRMIPScore  SNPsInPeak    PeakStart      PeakEnd  \\\n",
       "0       2,882,752                 6           1    2,882,752    2,882,752   \n",
       "1      17,966,847                 7           1   17,966,847   17,966,847   \n",
       "2      80,602,342                 5           1   80,602,342   80,602,342   \n",
       "3     104,530,524                17           1  104,530,524  104,530,524   \n",
       "4     241,078,399                 5           1  241,078,399  241,078,399   \n",
       "\n",
       "   PeakLength SecondClosestGeneToPeak ThirdClosestGeneToPeak  \\\n",
       "0           0          Zm00001d027321         Zm00001d027319   \n",
       "1           0          Zm00001d027934         Zm00001d027935   \n",
       "2           0          Zm00001d029629         Zm00001d029627   \n",
       "3           0          Zm00001d030087         Zm00001d030088   \n",
       "4           0          Zm00001d032874         Zm00001d032870   \n",
       "\n",
       "   NumberOfSignificantTraits    AllSignificantTraits  \\\n",
       "0                          1  BushelAcreEquivalent_J   \n",
       "1                          1   NodesWithBraceRoots_J   \n",
       "2                          1              EarWidth_J   \n",
       "3                          1          EarsPerPlant_J   \n",
       "4                          1             RootAngle_I   \n",
       "\n",
       "   NumberOfSignificantPhenotypeGroups PhenotypeGroups ClosestCore  \n",
       "0                                   1       Agronomic         Yes  \n",
       "1                                   1            Root         Yes  \n",
       "2                                   1   Inflorescence         Yes  \n",
       "3                                   1       Agronomic         Yes  \n",
       "4                                   1            Root         Yes  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_cdf[\"ClosestCore\"] = \"Yes\"\n",
    "gwas_cdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d70abf1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the same for second- and third-closest genes\n",
    "gwas_scdf = scdf.merge(gwas)\n",
    "gwas_tcdf = tcdf.merge(gwas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1467414d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "101"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gwas_cdf[\"peakName\"].unique())+len(gwas_scdf[\"peakName\"].unique())+len(gwas_tcdf[\"peakName\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3e4149bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many unique peaks does this represent?\n",
    "cp = list(gwas_cdf[\"peakName\"].unique())\n",
    "sp = list(gwas_scdf[\"peakName\"].unique())\n",
    "tp = list(gwas_tcdf[\"peakName\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "95b38aaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "97"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(cp+sp+tp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e1fa1c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save this set as a dataframe\n",
    "cg_gwas_hits_names = list(set(cp+sp+tp))\n",
    "cg_gwas_hits = gwas[gwas[\"peakName\"].isin(cg_gwas_hits_names)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "837876d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peakName</th>\n",
       "      <th>PeakChromosome</th>\n",
       "      <th>PeakSNPPosition</th>\n",
       "      <th>MaximumRMIPScore</th>\n",
       "      <th>SNPsInPeak</th>\n",
       "      <th>PeakStart</th>\n",
       "      <th>PeakEnd</th>\n",
       "      <th>PeakLength</th>\n",
       "      <th>ClosestGeneToPeak</th>\n",
       "      <th>SecondClosestGeneToPeak</th>\n",
       "      <th>ThirdClosestGeneToPeak</th>\n",
       "      <th>NumberOfSignificantTraits</th>\n",
       "      <th>AllSignificantTraits</th>\n",
       "      <th>NumberOfSignificantPhenotypeGroups</th>\n",
       "      <th>PhenotypeGroups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>280,994,311</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>280,994,269</td>\n",
       "      <td>281,039,919</td>\n",
       "      <td>45650</td>\n",
       "      <td>Zm00001d034005</td>\n",
       "      <td>Zm00001d034006</td>\n",
       "      <td>Zm00001d034007</td>\n",
       "      <td>2</td>\n",
       "      <td>EarHeight_M;RootAngle2_O</td>\n",
       "      <td>2</td>\n",
       "      <td>Root;Vegetative</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "      <td>297,580,977</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>297,580,977</td>\n",
       "      <td>297,582,588</td>\n",
       "      <td>1611</td>\n",
       "      <td>Zm00001d034600</td>\n",
       "      <td>Zm00001d034601</td>\n",
       "      <td>Zm00001d034602</td>\n",
       "      <td>2</td>\n",
       "      <td>BranchNumber1_C;BranchZoneLength_J</td>\n",
       "      <td>1</td>\n",
       "      <td>Inflorescence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>65</td>\n",
       "      <td>10</td>\n",
       "      <td>149,988,757</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>149,958,731</td>\n",
       "      <td>149,988,757</td>\n",
       "      <td>30026</td>\n",
       "      <td>Zm00001d026689</td>\n",
       "      <td>Zm00001d026687</td>\n",
       "      <td>Zm00001d026690</td>\n",
       "      <td>1</td>\n",
       "      <td>CobWeightGrams_J</td>\n",
       "      <td>1</td>\n",
       "      <td>Agronomic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>86</td>\n",
       "      <td>3</td>\n",
       "      <td>230,172,221</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>230,171,635</td>\n",
       "      <td>230,172,221</td>\n",
       "      <td>586</td>\n",
       "      <td>Zm00001d044506</td>\n",
       "      <td>Zm00001d044507</td>\n",
       "      <td>Zm00001d044504</td>\n",
       "      <td>1</td>\n",
       "      <td>SAMVolume_M</td>\n",
       "      <td>1</td>\n",
       "      <td>Cellular/Biochemical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>93</td>\n",
       "      <td>4</td>\n",
       "      <td>190,410,017</td>\n",
       "      <td>26</td>\n",
       "      <td>6</td>\n",
       "      <td>190,332,784</td>\n",
       "      <td>190,410,017</td>\n",
       "      <td>77233</td>\n",
       "      <td>Zm00001d052457</td>\n",
       "      <td>Zm00001d052458</td>\n",
       "      <td>Zm00001d052453</td>\n",
       "      <td>6</td>\n",
       "      <td>CrudeAsh_K;EarHeight_L;Ncombustion_K;Nkjeltec_...</td>\n",
       "      <td>2</td>\n",
       "      <td>SeedComposition;Vegetative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    peakName  PeakChromosome PeakSNPPosition  MaximumRMIPScore  SNPsInPeak  \\\n",
       "24        25               1     280,994,311                13           3   \n",
       "55        56               1     297,580,977                12           2   \n",
       "64        65              10     149,988,757                10           2   \n",
       "85        86               3     230,172,221                 8           2   \n",
       "92        93               4     190,410,017                26           6   \n",
       "\n",
       "      PeakStart      PeakEnd  PeakLength ClosestGeneToPeak  \\\n",
       "24  280,994,269  281,039,919       45650    Zm00001d034005   \n",
       "55  297,580,977  297,582,588        1611    Zm00001d034600   \n",
       "64  149,958,731  149,988,757       30026    Zm00001d026689   \n",
       "85  230,171,635  230,172,221         586    Zm00001d044506   \n",
       "92  190,332,784  190,410,017       77233    Zm00001d052457   \n",
       "\n",
       "   SecondClosestGeneToPeak ThirdClosestGeneToPeak  NumberOfSignificantTraits  \\\n",
       "24          Zm00001d034006         Zm00001d034007                          2   \n",
       "55          Zm00001d034601         Zm00001d034602                          2   \n",
       "64          Zm00001d026687         Zm00001d026690                          1   \n",
       "85          Zm00001d044507         Zm00001d044504                          1   \n",
       "92          Zm00001d052458         Zm00001d052453                          6   \n",
       "\n",
       "                                 AllSignificantTraits  \\\n",
       "24                           EarHeight_M;RootAngle2_O   \n",
       "55                 BranchNumber1_C;BranchZoneLength_J   \n",
       "64                                   CobWeightGrams_J   \n",
       "85                                        SAMVolume_M   \n",
       "92  CrudeAsh_K;EarHeight_L;Ncombustion_K;Nkjeltec_...   \n",
       "\n",
       "    NumberOfSignificantPhenotypeGroups             PhenotypeGroups  \n",
       "24                                   2             Root;Vegetative  \n",
       "55                                   1               Inflorescence  \n",
       "64                                   1                   Agronomic  \n",
       "85                                   1        Cellular/Biochemical  \n",
       "92                                   2  SeedComposition;Vegetative  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cg_gwas_hits.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8141831a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['EarHeight_M;RootAngle2_O', 'BranchNumber1_C;BranchZoneLength_J',\n",
       "       'CobWeightGrams_J', 'SAMVolume_M',\n",
       "       'CrudeAsh_K;EarHeight_L;Ncombustion_K;Nkjeltec_K;PlantHeight_L;Protein_K',\n",
       "       'SMV10DAI_B;SMV14DAI_B;SMVAUPDC_B', 'EarsPerPlant_J',\n",
       "       'TillersPerPlant_J', 'BranchDensity_C', 'NodesWithBraceRoots_J',\n",
       "       'BranchesPerTassel_J', 'RootAngle_I', 'RootDepth2_O', 'EarWidth_J',\n",
       "       'Ncombustion_K;Nkjeltec_K', 'Protein_K', 'EarHeight_M',\n",
       "       'BiomassYield_G', 'BushelAcreEquivalent_J', 'Fructose_K',\n",
       "       'PH-EH_L', 'Nodes_M',\n",
       "       'SMV10DAI_B;SMV14DAI_B;SMV21DAI_B;SMV28DAI_B',\n",
       "       'TasselLengthError_P', 'PlantHeight_J', 'SMV7DAI_B', 'SMV28DAI_B',\n",
       "       'senescence2_F', 'BranchNumber_P', 'KernelRowNumber_J',\n",
       "       'LeafCuticularConductance7_H', 'SMVAUPDC_B',\n",
       "       'SMV28DAI_B;SMV35DAI_B;SMVAUPDC_B', 'Anthesis_G',\n",
       "       'KernelsPerRow_J', 'Starch_K', 'BraceRoot_N',\n",
       "       'RootAngle1_O;RootAngle2_O', 'StalkDiamThick_N;peri_N',\n",
       "       'LeafAreaIndex_J', 'StalkDiamThin_N', 'BranchZoneLength_C',\n",
       "       'BranchNumber1_C', 'RootArea2_O;RootArea4_O', 'EHdivPH_L',\n",
       "       'FractalDimension_C', 'Toruosity_C', 'VascularBundleDensity_D',\n",
       "       'Glucose_K', 'BranchDensity_C;BranchNumber1_C;BranchNumber2_C',\n",
       "       'Anthesis7_H', 'Anthesis6_H', 'StalkLodgingPct_J',\n",
       "       'SMV10DAI_B;SMV14DAI_B', 'Ash_K;Nkjeltec_K;Protein_K',\n",
       "       'senescence3_F', 'EarFilledLength_J',\n",
       "       'SouthernRustSeverityScore_J', 'CentralSpikeLengthError_P',\n",
       "       'StalkDiameter_D', 'LeafAngle_J', 'LeafWidth_J', 'SAMVolume1_N',\n",
       "       'BranchZoneLengthGT_P', 'Ash_K', 'Lignin_N', 'RootWidth1_O',\n",
       "       'LowestBranchAngleGT_P', 'GrainPercentMoisture_J', 'Compactness_C',\n",
       "       'Fiber_K', 'AnkomCrudeFiber_K', 'AnkomCrudeFiber_K;Ash_K',\n",
       "       'CentralSpikeLengthGT_P'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cg_gwas_hits[\"AllSignificantTraits\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "715f8f38",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "cdac98d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Zm00001eb000010</th>\n",
       "      <th>Zm00001eb000020</th>\n",
       "      <th>Zm00001eb000050</th>\n",
       "      <th>Zm00001eb000060</th>\n",
       "      <th>Zm00001eb000070</th>\n",
       "      <th>Zm00001eb000080</th>\n",
       "      <th>Zm00001eb000100</th>\n",
       "      <th>...</th>\n",
       "      <th>Zm00001eb442810</th>\n",
       "      <th>Zm00001eb442820</th>\n",
       "      <th>Zm00001eb442840</th>\n",
       "      <th>Zm00001eb442850</th>\n",
       "      <th>Zm00001eb442870</th>\n",
       "      <th>Zm00001eb442890</th>\n",
       "      <th>Zm00001eb442910</th>\n",
       "      <th>Zm00001eb442960</th>\n",
       "      <th>Zm00001eb442980</th>\n",
       "      <th>Zm00001eb443030</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11933261</td>\n",
       "      <td>PRJNA637522</td>\n",
       "      <td>Drought</td>\n",
       "      <td>12.553818</td>\n",
       "      <td>2.321077</td>\n",
       "      <td>0.042520</td>\n",
       "      <td>12.932676</td>\n",
       "      <td>5.253755</td>\n",
       "      <td>11.105837</td>\n",
       "      <td>0.409268</td>\n",
       "      <td>...</td>\n",
       "      <td>0.171184</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.309501</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11933272</td>\n",
       "      <td>PRJNA637522</td>\n",
       "      <td>Drought</td>\n",
       "      <td>16.255838</td>\n",
       "      <td>3.110372</td>\n",
       "      <td>0.405226</td>\n",
       "      <td>7.214039</td>\n",
       "      <td>1.902461</td>\n",
       "      <td>2.346186</td>\n",
       "      <td>0.170305</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108052</td>\n",
       "      <td>0.127878</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.703281</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11933250</td>\n",
       "      <td>PRJNA637522</td>\n",
       "      <td>Drought</td>\n",
       "      <td>9.028815</td>\n",
       "      <td>2.984479</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.092442</td>\n",
       "      <td>2.586555</td>\n",
       "      <td>16.186141</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.417565</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.254123</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.213349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11933029</td>\n",
       "      <td>PRJNA637522</td>\n",
       "      <td>Control</td>\n",
       "      <td>8.201340</td>\n",
       "      <td>2.385748</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.726808</td>\n",
       "      <td>1.926412</td>\n",
       "      <td>19.600487</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.370075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11933040</td>\n",
       "      <td>PRJNA637522</td>\n",
       "      <td>Drought</td>\n",
       "      <td>10.371251</td>\n",
       "      <td>2.799099</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.280629</td>\n",
       "      <td>3.771234</td>\n",
       "      <td>19.717683</td>\n",
       "      <td>0.143764</td>\n",
       "      <td>...</td>\n",
       "      <td>0.178304</td>\n",
       "      <td>0.012158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.625225</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.352959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 39007 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Sample   BioProject Treatment  Zm00001eb000010  Zm00001eb000020  \\\n",
       "0  SRR11933261  PRJNA637522   Drought        12.553818         2.321077   \n",
       "1  SRR11933272  PRJNA637522   Drought        16.255838         3.110372   \n",
       "2  SRR11933250  PRJNA637522   Drought         9.028815         2.984479   \n",
       "3  SRR11933029  PRJNA637522   Control         8.201340         2.385748   \n",
       "4  SRR11933040  PRJNA637522   Drought        10.371251         2.799099   \n",
       "\n",
       "   Zm00001eb000050  Zm00001eb000060  Zm00001eb000070  Zm00001eb000080  \\\n",
       "0         0.042520        12.932676         5.253755        11.105837   \n",
       "1         0.405226         7.214039         1.902461         2.346186   \n",
       "2         0.000000         3.092442         2.586555        16.186141   \n",
       "3         0.000000         1.726808         1.926412        19.600487   \n",
       "4         0.000000         1.280629         3.771234        19.717683   \n",
       "\n",
       "   Zm00001eb000100  ...  Zm00001eb442810  Zm00001eb442820  Zm00001eb442840  \\\n",
       "0         0.409268  ...         0.171184         0.000000              0.0   \n",
       "1         0.170305  ...         0.108052         0.127878              0.0   \n",
       "2         0.000000  ...         0.000000         0.000000              0.0   \n",
       "3         0.000000  ...         0.000000         0.000000              0.0   \n",
       "4         0.143764  ...         0.178304         0.012158              0.0   \n",
       "\n",
       "   Zm00001eb442850  Zm00001eb442870  Zm00001eb442890  Zm00001eb442910  \\\n",
       "0              0.0              0.0         0.309501              0.0   \n",
       "1              0.0              0.0         6.703281              0.0   \n",
       "2              0.0              0.0         0.417565              0.0   \n",
       "3              0.0              0.0         0.000000              0.0   \n",
       "4              0.0              0.0         9.625225              0.0   \n",
       "\n",
       "   Zm00001eb442960  Zm00001eb442980  Zm00001eb443030  \n",
       "0         0.000000              0.0         0.000000  \n",
       "1         0.000000              0.0         0.000000  \n",
       "2         0.254123              0.0         1.213349  \n",
       "3         0.000000              0.0         1.370075  \n",
       "4         0.000000              0.0         2.352959  \n",
       "\n",
       "[5 rows x 39007 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# start setup from scratch\n",
    "# I need a list of all the genes, and I need to note whether they are core or not, and whether they are close to GWAS hits or not\n",
    "\n",
    "# load tpm data with list of genes\n",
    "tpm = pd.read_csv(\"../../data/rawtpm_bptreat_noPEG.tsv\",sep=\"\\t\",header=\"infer\")\n",
    "tpm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b0157cf0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Sample</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>SRR11933261</th>\n",
       "      <th>SRR11933272</th>\n",
       "      <th>SRR11933250</th>\n",
       "      <th>SRR11933029</th>\n",
       "      <th>SRR11933040</th>\n",
       "      <th>SRR11932822</th>\n",
       "      <th>SRR11932811</th>\n",
       "      <th>SRR11933230</th>\n",
       "      <th>SRR11932879</th>\n",
       "      <th>...</th>\n",
       "      <th>Ms71D3C</th>\n",
       "      <th>Ki3D1C</th>\n",
       "      <th>CML228D1D</th>\n",
       "      <th>CML333D3D</th>\n",
       "      <th>MO18WD3C</th>\n",
       "      <th>B73D3C</th>\n",
       "      <th>NC358D3C</th>\n",
       "      <th>P39D3D</th>\n",
       "      <th>M162WD3D</th>\n",
       "      <th>M162WD1D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001eb000010</td>\n",
       "      <td>12.553818</td>\n",
       "      <td>16.255838</td>\n",
       "      <td>9.028815</td>\n",
       "      <td>8.201340</td>\n",
       "      <td>10.371251</td>\n",
       "      <td>37.430009</td>\n",
       "      <td>39.925873</td>\n",
       "      <td>30.677016</td>\n",
       "      <td>23.393003</td>\n",
       "      <td>...</td>\n",
       "      <td>1.417104</td>\n",
       "      <td>1.923525</td>\n",
       "      <td>1.427602</td>\n",
       "      <td>9.580153</td>\n",
       "      <td>1.228100</td>\n",
       "      <td>2.966207</td>\n",
       "      <td>1.791556</td>\n",
       "      <td>4.286976</td>\n",
       "      <td>3.435711</td>\n",
       "      <td>3.498243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001eb000020</td>\n",
       "      <td>2.321077</td>\n",
       "      <td>3.110372</td>\n",
       "      <td>2.984479</td>\n",
       "      <td>2.385748</td>\n",
       "      <td>2.799099</td>\n",
       "      <td>27.508819</td>\n",
       "      <td>22.440680</td>\n",
       "      <td>24.648455</td>\n",
       "      <td>7.595576</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.799671</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.925157</td>\n",
       "      <td>0.561768</td>\n",
       "      <td>0.176413</td>\n",
       "      <td>0.781353</td>\n",
       "      <td>0.379497</td>\n",
       "      <td>0.463832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001eb000050</td>\n",
       "      <td>0.042520</td>\n",
       "      <td>0.405226</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.304751</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001eb000060</td>\n",
       "      <td>12.932676</td>\n",
       "      <td>7.214039</td>\n",
       "      <td>3.092442</td>\n",
       "      <td>1.726808</td>\n",
       "      <td>1.280629</td>\n",
       "      <td>29.510498</td>\n",
       "      <td>22.148225</td>\n",
       "      <td>22.170584</td>\n",
       "      <td>14.727189</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001eb000070</td>\n",
       "      <td>5.253755</td>\n",
       "      <td>1.902461</td>\n",
       "      <td>2.586555</td>\n",
       "      <td>1.926412</td>\n",
       "      <td>3.771234</td>\n",
       "      <td>7.005587</td>\n",
       "      <td>7.590336</td>\n",
       "      <td>5.274585</td>\n",
       "      <td>2.177748</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.451827</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.018369</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.660372</td>\n",
       "      <td>0.748587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1982 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Sample           GeneID  SRR11933261  SRR11933272  SRR11933250  SRR11933029  \\\n",
       "0       Zm00001eb000010    12.553818    16.255838     9.028815     8.201340   \n",
       "1       Zm00001eb000020     2.321077     3.110372     2.984479     2.385748   \n",
       "2       Zm00001eb000050     0.042520     0.405226     0.000000     0.000000   \n",
       "3       Zm00001eb000060    12.932676     7.214039     3.092442     1.726808   \n",
       "4       Zm00001eb000070     5.253755     1.902461     2.586555     1.926412   \n",
       "\n",
       "Sample  SRR11933040  SRR11932822  SRR11932811  SRR11933230  SRR11932879  ...  \\\n",
       "0         10.371251    37.430009    39.925873    30.677016    23.393003  ...   \n",
       "1          2.799099    27.508819    22.440680    24.648455     7.595576  ...   \n",
       "2          0.000000     0.000000     0.000000     0.000000     0.304751  ...   \n",
       "3          1.280629    29.510498    22.148225    22.170584    14.727189  ...   \n",
       "4          3.771234     7.005587     7.590336     5.274585     2.177748  ...   \n",
       "\n",
       "Sample   Ms71D3C    Ki3D1C  CML228D1D  CML333D3D  MO18WD3C    B73D3C  \\\n",
       "0       1.417104  1.923525   1.427602   9.580153  1.228100  2.966207   \n",
       "1       0.000000  1.799671   0.000000   0.000000  1.925157  0.561768   \n",
       "2       0.000000  0.000000   0.000000   0.000000  0.000000  0.000000   \n",
       "3       0.000000  0.000000   0.000000   0.000000  0.000000  0.000000   \n",
       "4       0.000000  0.451827   0.000000   1.018369  0.000000  0.000000   \n",
       "\n",
       "Sample  NC358D3C    P39D3D  M162WD3D  M162WD1D  \n",
       "0       1.791556  4.286976  3.435711  3.498243  \n",
       "1       0.176413  0.781353  0.379497  0.463832  \n",
       "2       0.000000  0.000000  0.000000  0.000000  \n",
       "3       0.000000  0.000000  0.000000  0.000000  \n",
       "4       0.000000  0.000000  1.660372  0.748587  \n",
       "\n",
       "[5 rows x 1982 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# wrangle tpm data\n",
    "ttpm = tpm.set_index(\"Sample\").drop([\"BioProject\",\"Treatment\"],axis=1).transpose().reset_index().rename(columns={\"index\":\"GeneID\"})\n",
    "ttpm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96cec41f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cgdict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "10876fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# isolate the gene IDs\n",
    "gid = list(ttpm[\"GeneID\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "605d2ca7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Requested ID</th>\n",
       "      <th>Matched ID(s)</th>\n",
       "      <th>Releases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Zm00001d052837</td>\n",
       "      <td>Zm00001eb199410</td>\n",
       "      <td>110: Zm00001eb199410.1104: Zm00001eb199410.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Zm00001d025912</td>\n",
       "      <td>Zm00001eb427160</td>\n",
       "      <td>110: Zm00001eb427160.1104: Zm00001eb427160.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Zm00001d008485</td>\n",
       "      <td>Zm00001eb334630</td>\n",
       "      <td>110: Zm00001eb334630.1104: Zm00001eb334630.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Zm00001d033568</td>\n",
       "      <td>Zm00001eb053580</td>\n",
       "      <td>110: Zm00001eb053580.1104: Zm00001eb053580.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zm00001d012957</td>\n",
       "      <td>Zm00001eb211650</td>\n",
       "      <td>110: Zm00001eb211650.1104: Zm00001eb211650.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Requested ID    Matched ID(s)  \\\n",
       "0  Zm00001d052837  Zm00001eb199410   \n",
       "1  Zm00001d025912  Zm00001eb427160   \n",
       "2  Zm00001d008485  Zm00001eb334630   \n",
       "3  Zm00001d033568  Zm00001eb053580   \n",
       "4  Zm00001d012957  Zm00001eb211650   \n",
       "\n",
       "                                       Releases  \n",
       "0  110: Zm00001eb199410.1104: Zm00001eb199410.1  \n",
       "1  110: Zm00001eb427160.1104: Zm00001eb427160.1  \n",
       "2  110: Zm00001eb334630.1104: Zm00001eb334630.1  \n",
       "3  110: Zm00001eb053580.1104: Zm00001eb053580.1  \n",
       "4  110: Zm00001eb211650.1104: Zm00001eb211650.1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load the results from the ID History Converter from Ensembl Plants\n",
    "idres = pd.read_csv(\"../../data/Results-Zea_mays_Tools_IDMapper_.csv\",sep=\",\",header=\"infer\")\n",
    "idres.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e19baf56",
   "metadata": {},
   "outputs": [],
   "source": [
    "# covert to a dictionary with V5 as key and V4 as value\n",
    "gwasdict = {}\n",
    "for i in range(len(idres.index)):\n",
    "    k = idres.iloc[i,1]\n",
    "    gwasdict[k] = idres.iloc[i,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7024e201",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this and the next two cells are for testing association with all genes associated with GWAS hits - first, second, and third\n",
    "## closest\n",
    "iscore = []\n",
    "isgwas = []\n",
    "for g in gid:\n",
    "    if g in list(cgdict.keys()):\n",
    "        iscore.append(\"Y\")\n",
    "    else:\n",
    "        iscore.append(\"N\")\n",
    "    if g in list(idres[\"Matched ID(s)\"].unique()):\n",
    "        isgwas.append(\"Y\")\n",
    "    else:\n",
    "        isgwas.append(\"N\")\n",
    "        \n",
    "# make dataframe\n",
    "gdf = pd.DataFrame(list(zip(gid,iscore,isgwas)),columns=[\"GeneID\",\"isCore\",\"isGWAShit\"])\n",
    "gdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a79ec9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a crosstab - contingency table\n",
    "data = pd.crosstab(index=gdf[\"isGWAShit\"],columns=gdf[\"isCore\"])\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1efe12a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run Fisher's exact test\n",
    "odds_ratio, p_value = stats.fisher_exact(data)\n",
    "print('odd ratio is : ' + str(odds_ratio))\n",
    "print('p value is : ' + str(p_value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bab988eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['peakName', 'PeakChromosome', 'PeakSNPPosition', 'MaximumRMIPScore',\n",
       "       'SNPsInPeak', 'PeakStart', 'PeakEnd', 'PeakLength', 'ClosestGeneToPeak',\n",
       "       'SecondClosestGeneToPeak', 'ThirdClosestGeneToPeak',\n",
       "       'NumberOfSignificantTraits', 'AllSignificantTraits',\n",
       "       'NumberOfSignificantPhenotypeGroups', 'PhenotypeGroups'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e3e8dc70",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test for association with first closest genes and core genes\n",
    "iscore = []\n",
    "isgwas = []\n",
    "for g in gid:\n",
    "    if g in list(cgdict.keys()):\n",
    "        iscore.append(\"Y\")\n",
    "    else:\n",
    "        iscore.append(\"N\")\n",
    "        \n",
    "    if g in gwasdict.keys():\n",
    "        gv4 = gwasdict[g]\n",
    "        if gv4 in list(gwas[\"ClosestGeneToPeak\"].unique()):\n",
    "            isgwas.append(\"Y\")\n",
    "        else:\n",
    "            isgwas.append(\"N\")\n",
    "    else:\n",
    "        isgwas.append(\"N\")\n",
    "gdf = pd.DataFrame(list(zip(gid,iscore,isgwas)),columns=[\"GeneID\",\"isCore\",\"isFirstGWAS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a503729f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>isCore</th>\n",
       "      <th>N</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isFirstGWAS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>36949</td>\n",
       "      <td>831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>1196</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "isCore           N    Y\n",
       "isFirstGWAS            \n",
       "N            36949  831\n",
       "Y             1196   28"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make a crosstab - contingency table\n",
    "data = pd.crosstab(index=gdf[\"isFirstGWAS\"],columns=gdf[\"isCore\"])\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "41c893f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "odd ratio is : 1.0409467579456593\n",
      "p value is : 0.842744972159088\n"
     ]
    }
   ],
   "source": [
    "# run Fisher's exact test\n",
    "odds_ratio, p_value = stats.fisher_exact(data)\n",
    "print('odd ratio is : ' + str(odds_ratio))\n",
    "print('p value is : ' + str(p_value))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0b6a8a0",
   "metadata": {},
   "source": [
    "## Testing for association with traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b767e99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a list of the traits\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
